### PR TITLE
Gameplay logic, tests, db updates

### PR DESCRIPTION
**Pull request by rmarincola**

### CODE REVIEW

File: cardtel/utils/gameplay.py [hunk 1]
```diff
@@ -1,11 +1,54 @@
...
+        return Player.objects.create(game=game, user=user, play_order=0)
+
+    play_order = Player.objects.filter(
+        game=game
+    ).order_by(
+        '-play_order'
+    ).values_list(
+        'play_order',
+        flat=True
+    ).first() + 1
```
Comment (bweems23) [inline cardtel/utils/gameplay.py]: should this be ranked by user score?

File: cardtel/utils/gameplay.py [hunk 2]
```diff
@@ -1,11 +1,54 @@
...
+
+    return Player.objects.create(game=game, user=user, play_order=play_order)
+
+def initialize_game(game):
+    # Set current turn to the player with the lowest overall score
```
Comment (bweems23) [inline cardtel/utils/gameplay.py]: move down a line

File: cardtel/utils/gameplay.py [hunk 3]
```diff
@@ -1,11 +1,54 @@
...
-	raise NotImplementedError
+    player = Player.objects.filter(game=game, user=user).first()
+    player.delete()
 
 def update_user_scores(game):
-	raise NotImplementedError
+    players = Player.objects.filter(game=game).all()
+    for player in players:
+        user = player.user
+        user.score = user.score + player.score
```
Comment (bweems23) [inline cardtel/utils/gameplay.py]: no silly! that's not how you update a user's score!

File: cardtel/models.py [hunk 4]
```diff
@@ -33,21 +34,44 @@ def point_is_over(self):
...
         db_table = 'player'
         app_label = 'cardtel'
         unique_together = ('game', 'user')
 
+    created_at = models.DateTimeField(
+        auto_now_add=True,
+    )
     game = models.ForeignKey(
-        'cardtel.Game',
+        'cardtel.Game',
```
Comment (bweems23) [inline cardtel/models.py]: :(

File: cardtel/models.py [hunk 5]
```diff
@@ -33,21 +34,44 @@ def point_is_over(self):
...
+        app_label = 'cardtel'
+
+    created_at = models.DateTimeField(
+        auto_now_add=True,
+    )
+    winner = models.ForeignKey(
+        'cardtel.Player',
+        related_name='points',
+    )
+    winning_cards = models.ManyToManyField(
```
Comment (bweems23) [inline cardtel/models.py:53]: will this be the whole table or just the cards that won?

File: cardtel/utils/gameplay.py [hunk 6]
```diff
@@ -1,11 +1,108 @@
...
+        'play_order',
+        flat=True
+    ).first() + 1
+
+    return Player.objects.create(game=game, user=user, play_order=play_order)
+
+def initialize_game(game):
+    """
+    Set current turn to the player with the lowest overall score
+    """
```
Comment (bweems23) [inline cardtel/utils/gameplay.py:41]: might as well add these comments to the other functions

File: cardtel/models.py (hunk 5 above)
Comment (rmarincola) [reply cardtel/models.py:53]: i was thinking the cards that won

File: cardtel/utils/gameplay.py [hunk 7]
```diff
@@ -1,11 +1,108 @@
...
+
+    return ranking_order
+
+def assign_final_points_to_players(ranked_players):
+    """
+    Given a final ranking of players, assign points to
+    the associated user's overall score.
+    """
+    num_players = len(ranked_players)
+    points_awarded = num_players / 2
```
Comment (bweems23) [inline cardtel/utils/gameplay.py]: max_points_awarded

Comment (bweems23) [reply cardtel/utils/gameplay.py]: or comment that this is the max for readability

File: cardtel/utils/tests/test_game.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        assert player.play_order == 1
+
+    def test_initialize_game(self):
+        game = self.game
+        user = self.user
+        player = add_player_to_game(game, user)
+        game = initialize_game(game)
+        assert game.current_turn == player
+
+        other_user = User.objects.create(username='bweems')
```
Comment (bweems23) [inline cardtel/utils/tests/test_game.py]: I'd probably add this player to the game before the previous assertion to confirm that the order is actually being flipped by the score change

File: cardtel/utils/tests/test_game.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+        Point.objects.create(winner=player)
+        Point.objects.create(winner=other_player)
+        Point.objects.create(winner=third_player)
+
+        game = update_user_scores(game)
+        player = Player.objects.get(id=player.id)
+        assert player.user.score == -1
+        other_player = Player.objects.get(id=other_player.id)
+        assert other_player.user.score == 1
```
Comment (bweems23) [inline cardtel/utils/tests/test_game.py:122]: I am forever winning test cardtel!

**Comment by bweems23** [comment]
LGTM nice tests!

**Comment by rmarincola** [comment]
@bweems23 updated based on comments if you wanna take another look!

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+sudo: false
```
Comment (bweems23) [inline .travis.yml:1]: Nice!!!

File: cardtel/utils/gameplay.py [hunk 11]
```diff
@@ -1,11 +1,141 @@
...
+    if player_one.score < player_two.score:
+        return 1
+    if player_one.score > player_two.score:
+        return -1
+
+    player_one_last_point = player_one.points.last()
+    player_two_last_point = player_two.points.last()
+
+    if not player_one_last_point and not player_two_last_point:
+        return 0
```
Comment (bweems23) [inline cardtel/utils/gameplay.py]: call rand on list w/[-1, 1]

**Comment by bweems23** [comment]
lgtm

**Pull request merged**